didman/didman.go

```diff
@@ -0,0 +1,153 @@
+	"github.com/nuts-foundation/go-did/did"
+	"github.com/nuts-foundation/nuts-node/vdr/types"
+	"github.com/shengdoushi/base58"
+)
+
+// ModuleName contains the name of this module: Didman
+const ModuleName = "Didman"
+
+// ErrDuplicateService is returned when a DID Document already contains a service for the given type
+var ErrDuplicateService = errors.New("service type already defined")
+
+// ErrServiceInUse is returned when a service is deleted but in use by other services
+var ErrServiceInUse = errors.New("service is referenced by 1 or more services")
+
+// ErrServiceNotFound is returned when the service is not found on a DID
+var ErrServiceNotFound = errors.New("service not found in DID Document")
+
+type didman struct {
+	docResolver types.DocResolver
+	store       types.Store
+	vdr         types.VDR
+}
+
+// NewDidmanInstance creates a new didman instance with services set
+func NewDidmanInstance(docResolver types.DocResolver, store types.Store, vdr types.VDR) Didman {
+	return &didman{
+		docResolver: docResolver,
+		store:       store,
+		vdr:         vdr,
+	}
+}
+
+func (d *didman) Name() string {
+	return ModuleName
+}
+
+func (d *didman) AddEndpoint(id did.DID, serviceType string, u url.URL) error {
+	doc, meta, err := d.docResolver.Resolve(id, nil)
+	if err != nil {
+		return err
+	}
+
+	// check for duplicate service type
+	for _, s := range doc.Service {
+		if s.Type == serviceType {
+			return ErrDuplicateService
+		}
+	}
+
+	// construct service with correct ID
+	service := constructService(id, serviceType, u)
+	doc.Service = append(doc.Service, service)
+
+	return d.vdr.Update(id, meta.Hash, *doc, nil)
+}
+
+func (d *didman) DeleteService(serviceID ssi.URI) error {
+	id, err := did.ParseDID(serviceID.String())
+	if err != nil {
+		return err
+	}
+	id.Fragment = ""
+
+	doc, meta, err := d.docResolver.Resolve(*id, nil)
+	if err != nil {
+		return err
+	}
+
+	// check for existing use
+	if err = d.store.Iterate(func(doc did.Document, metadata types.DocumentMetadata) error {
+		if referencesService(doc, serviceID) {
+			return ErrServiceInUse
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	// remove service
+	j := 0
+	for _, s := range doc.Service {
+		if s.ID != serviceID {
+			doc.Service[j] = s
+			j++
+		}
+	}
+	doc.Service = doc.Service[:j]
+
+	return d.vdr.Update(*id, meta.Hash, *doc, nil)
+}
+
+func constructService(id did.DID, serviceType string, u url.URL) did.Service {
+	service := did.Service{
+		Type:            serviceType,
+		ServiceEndpoint: u.String(),
+	}
+
+	service.ID = ssi.URI{}
+
+	bytes, _ := json.Marshal(service)
+	shaBytes := sha256.Sum256(bytes)
+	d := id.URI()
+	d.Fragment = base58.Encode(shaBytes[:], base58.BitcoinAlphabet)
+	service.ID = d
+
+	return service
+}
+
+type compoundService map[string]string
+
+func referencesService(doc did.Document, serviceID ssi.URI) bool {
+	id := serviceID.String()
+	for _, s := range doc.Service {
+		cs := compoundService{}
```

Comment:
copy/paste from `auth/services/util.go:48`?

Reply:
Ahhh the `type`. Maybe I can move it to the VDR as public type.